### ISSUE TITLE
Variable explorer is broken in 2.0.0beta6

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#434]
_From [DavidAnt...@gmail.com](https://code.google.com/u/108698404730356683634/) on 2010-11-01T17:43:40Z_

What steps will reproduce the problem?

 1. Run a program either in existing ipython console or new python console
2. Double click on any of the variables to edit them

Attached is the observed output.  I was previously running revision 939 (deeed3f88f2d) and did not see this problem until updating my hg repository to 2.0.0beta6

**Attachment:** [variable_explorer.PNG](http://code.google.com/p/spyderlib/issues/detail?id=434)

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=434_

**Comment by spyder-issues-migrator** [comment]
_From [DavidAnt...@gmail.com](https://code.google.com/u/108698404730356683634/) on 2010-11-01T15:59:29Z_

Repeatability of this bug is a little haphazard.  However I can confirm that it still occurs after running spyder.py --reset.  It may also be linked to the fact that a matplotlib figure is open at the same time, but hard to be sure.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-02T01:00:11Z_

Would it be possible for you to write a test script allowing to reproduce this bug?
Because I really can't reproduce it...

**Status:** Accepted  
  **Labels:** -Priority-Medium Priority-High Milestone-Release2.0

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-02T01:11:36Z_

Another thing:

FYI, the dictionary which is opened for edition (instead of the variable you selected) is the data sent by the external shell to Spyder (by the 'Monitor') to notify it about the external shell's namespace contents: variable types, sizes, etc. What is happening here is probably that when trying to get a variable from the external shell, Spyder receives instead this dictionary which has been sent by the monitor (at Spyder's request) when refreshing automatically the variable explorer. So my guess is that something went wrong during variable explorer refreshing process, and the communication was suspended so Spyder had never received the dictionary... or something like that if you follow me.

There may be a debug log file in '~/.spyder2' (on Windows: "C:\Document and Settings\your_username\.spyder2") named 'monitor.log'. According to the symptoms you described, this file should exist following this bug (unfortunately, doing 'python spyder.py --reset' remove all configuration files including the debug log files).

To debug this further, you may also enable the debug mode in spyderlib/widgets/externalshell/monitor.py: line 18, write "DEBUG = True" instead of "DEBUG = False". This should generate a log file containing all communications between the external shell and Spyder and named 'monitor_debug.log'.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-02T01:12:37Z_

I would also like to know what is your machine configuration: OS, PyQt version, Python version, ...

**Comment by spyder-issues-migrator** [comment]
_From [DavidAnt...@gmail.com](https://code.google.com/u/108698404730356683634/) on 2010-11-02T16:28:39Z_

Pierre,

I am running WinXP Pro 32, Python 2.6.5, PyQt 4.5.4

I have attached the monitor.log file

After rebooting my computer I am having a little trouble reproducing this bug at all.  I'll look into this further and keep you posted.

Thanks for all your hard work.

**Attachment:** [monitor.log](http://code.google.com/p/spyderlib/issues/detail?id=434#hc5)

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2010-11-02T20:50:32Z_

I can reproduce this behavior, but it is only under these circumstances:

1. It must be the first launch of Spyder after booting my machine. On any subsequent launch, I don't get the problem.
2. It only happens with some scripts. One script I can get to fail every time with a fresh OS and the first launch of Spyder. I can't share the full script, but here is the main module import list:

import numpy as np
import xlrd as xl
import matplotlib.pyplot as plt
import matplotlib.mlab as mplmlab
from enthought.mayavi import mlab

I have attached two error traces from the internal console. The first happens right after double clicking on a variable named "act_dep" to edit it in the variable explorer. The second happens after I hit "Cancel" from the variable editing pop-up. I have also attached the monitor_debug.log, too, but I don't know how useful it will be.

Hope this helps.

Spyder Hg revision 00ef524fe0c2 , Python 2.6.2, Qt 4.5.4, PyQt 4.5.4 on Win XP 32.

**Attachment:** [Internal Console after double click to edit a variable.txt Internal Console after Cancel.txt monitor_debug.log](http://code.google.com/p/spyderlib/issues/detail?id=434#hc6)

**Comment by spyder-issues-migrator** [comment]
_From [DavidAnt...@gmail.com](https://code.google.com/u/108698404730356683634/) on 2010-11-02T20:55:46Z_

I have chased down a few more details and can now reproduce this bug a little better

Firstly, I should mention that when this bug occurs, if I click on enough different variables in the variable explorer, then I can actually cause spyder to crash.

It seems to happen when I call a CPU intensive fortran routine which I have wrapped with f2py.  When running this program, I notice that the little timer on the console does not update every second, possibly meaning that this code is hogging the CPU and causing some events(?) to be missed somewhere.

The problem occurs if I have the variable explorer automatically updating while this code is running.  I followed your instructions and checked the file "monitor_debug.log".  I noticed that the entry '__get_global__(globals(), "full_kx")'
appears when I click on the variable "full_kx".  However, the contents of the log file are correct even when the variable explorer shows the nonsense entries, which correspond to a call to '__make_remote_view__(globals())'.

I debugged spyder a little, and put a breakpoint into the function monitor_get_global().  At this point I inspected the returned value, and it has the incorrect value.  I think the problem is with the functions read_packet() and write_packet() in monitor.py.  Somehow calls to read_packet() are getting missed, and all the previous responses are filling up the socket's receive buffer.  

The obvious solution would be to flush the socket, but looking at the python documentation, there doesn't seem to be any such functionality (I guess it doesn't exist in BSD sockets).  Maybe the best way to do it is for some kind of request number to be sent with the request for data.  This same number should then be returned with the answer.  The function read_packet() could then inspect this number, and discard any packets which do not match.

I guess it's not really possible to guarantee that no calls to read_packet() ever get missed.

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2010-11-02T21:26:07Z_

I can confirm that the script I mentioned above in Comment 6 that fails every time (at least with a clean OS and first launch of Spyder) contains some fairly intensive computations that could be hogging the CPU.

I can also confirm that the variable I'm trying to examine shows up correct in "monitor_debug.log" in my case, just as reported in Comment 7 above.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-03T03:20:52Z_

Thanks for all these explanations.

Just to be sure that I understand the symptoms:
When this bug occurs, you have the output shown in screenshot above:
  * only once: the first time you try to edit an object after calling the CPU intensive routine you were referring to?
  * every time you try to edit an object after calling this routine?

Could you try to remove the timeout on the read_packet function call in 'monitor_get_remote_view':
in 'spyderlib/widgets/externalshell/monitor.py', line 109:
replace:
    return read_packet(sock, timeout=5.)
by:
    return read_packet(sock)
and see what happens...

thanks!

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-03T05:26:36Z_

(BTW, I'm adding your names to the project contributors list, as you have been strongly implicated in Spyder's debugging process for a long time now)

**Comment by spyder-issues-migrator** [comment]
_From [jed.lud...@gmail.com](https://code.google.com/u/114333776555678317161/) on 2010-11-03T06:02:48Z_

In response to Comment 9, the behavior for me was exhibited every time I tried to edit an object. In other words, once the variable explorer was "broken" it remained broken for the rest of that Spyder session. Closing Spyder and re-launching would solve the issue for me.

I have removed the timeout as suggested in Comment 9. I have now gone through three successive reboot and test cycles, all of which have passed with success. Before removing the timeout I got it to fail essentially on demand by getting a fresh OS. Now it seems to be working.

Incidentally, I was informally watching the IPython console clock during these last three tests, and it appears that whatever process I'm running in my script is causing that clock to stop updating for 7 to 10 seconds. Whatever thread is trying to update the clock looks starved for that period. That might help explain why a 5 second packet read timeout could be the source of the original problem.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-03T07:58:07Z_

Yes that may be the reason why the socket receiving function timed out. 

The thread which is responsible to update the clock display is the main Spyder thread: but this more precisely the Qt event loop that is in charge of this. During a long process, the 'clock refresh' event will be put on hold in the event queue by Qt because this is not a high priority task for the Qt event loop (this is a QTimer, so it clearly depends on CPU occupancy).

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-03T10:51:42Z_

So, I propose two solutions:

(1) remove the timeout from 'monitor_get_remote_view'
(2) use the alternate monitor.py (attached file)

Pros/Cons:

(1) The main drawback of this solution is that if for some reason the remote process crash while constructing the remote view, Spyder will crash too because the 'read_packet' function is blocking.

(2) The 'monitor_get_remote_view' timeout is kept at 5 seconds, so if something goes wrong in the remote process during this (sometimes delicate) stage, Spyder won't crash: communication with monitor will simply be aborted. But, aborting the communication on one hand won't prevent the monitor from evaluating the command on the other hand, so from the monitor point of view, communication will continue. That's why, when communicating with the monitor, the function 'communicate' accept only packets corresponding to the command sent to the monitor. This is maybe *the* solution but it may block Spyder if a packet is lost for an unknown reason (in (1), if a packet is lost an error will (probably) occur but Spyder won't crash).

**Attachment:** [monitor.py](http://code.google.com/p/spyderlib/issues/detail?id=434#hc13)

**Comment by spyder-issues-migrator** [comment]
_From [DavidAnt...@gmail.com](https://code.google.com/u/108698404730356683634/) on 2010-11-03T16:18:26Z_

Regarding comment 9:

I also found that once the variable explorer got corrupted, it stayed that way.  Clicking on variables could bring up the output of __make_remote_view(), or the value of another variable, or if you clicked on the same variable enough times, then you can get the value of the correct variable.  Clicking on the variable explorer enough times always seems to crash spyder, with the error message in the attached image.

Getting rid of the timeout seemed to solve the problem for me

I am not sure which is the best solution to the problem.  I will test out the updated monitor.py to see if it ever fails.

**Attachment:** [spyder_crash.PNG](http://code.google.com/p/spyderlib/issues/detail?id=434#hc14)

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-04T15:08:50Z_

I'm about to commit a changeset that will automatically close this Issue and mark it as "Fixed". Please take it as a "assumed to be fixed": you are very welcome to test these changes on your respective configurations.

Thanks again for your help.

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2010-11-04T15:11:50Z_

This issue was closed by revision 1546188c9e .

**Status:** Fixed

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-04-24T07:36:45Z_

This issue was closed by revision 1546188c9e58 .

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-04-24T07:36:45Z_

This issue was closed by revision bfeb9bfe9cbe .
